text/0034-ternary-hash.md

```diff
@@ -0,0 +1,415 @@
+let mut kerl = Kerl::new();
+
+// `Kerl::digest` is a function that combines `Kerl::absorb` and `Kerl::squeeze`.
+// `Kerl::digest_into` combines `Kerl::absorb` with `Kerl::squeeze_into`. `CurlP` provides the same methods.
+let tx_hash = kerl.digest(&transaction);
+```
+
+The main proposal of this RFC are the `Sponge` trait and the `CurlP` and `Kerl` types that are implemented in terms of
+it. This RFC relies on the the presence of the types `TritBuf` and `Trits`, which are assumed to be owning and borrowing
+collections of binary-coded ternary in the `T1B1` encoding (one trit per byte), similar to
+[PathBuf](https://doc.rust-lang.org/std/path/struct.PathBuf.html) and
+[Path](https://doc.rust-lang.org/std/path/struct.Path.html).
+
+```rust
+/// The common interface of cryptographic hash functions that follow the sponge construction and that act on ternary.
+trait Sponge {
+    /// The expected length of the input to the sponge.
+    const IN_LEN: usize;
+
+    /// The length of the hash squeezed from the sponge.
+    const OUT_LEN: usize;
+
+    /// An error indicating a that a failure has occured during `absorb`.
+    type Error;
+
+    /// Absorb `input` into the sponge.
+    fn absorb(&mut self, input: &Trits) -> Result<(), Self::Error>;
+
+    /// Reset the inner state of the sponge.
+    fn reset(&mut self);
+
+    /// Squeeze the sponge into a buffer
+    fn squeeze_into(&mut self, buf: &mut Trits) -> Result<(), Self::Error>;
+
+    /// Convenience function using `Sponge::squeeze_into` to to return an owned version of the hash.
+    fn squeeze(&mut self) -> Result<TritBuf, Self::Error> {
+        let mut output = TritBuf::zeros(Self::OUT_LEN);
+        self.squeeze_into(&mut output)?;
+        Ok(output)
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge into a buffer, and reset the sponge in one go.
+    fn digest_into(&mut self, input: &Trits, buf: &mut Trits) -> Result<(), Self::Error> {
+        self.absorb(input)?;
+        self.squeeze_into(buf)?;
+        self.reset();
+        Ok(())
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge, and reset the sponge in one go.
+    /// Returns an owned versin of the hash.
+    fn digest(&mut self, input: &Trits) -> Result<TritBuf, Self::Error> {
+        self.absorb(input)?;
+        let output = self.squeeze()?;
+        self.reset();
+        Ok(output)
+    }
+}
+```
+
+Following the sponge metaphor, an input reference provided by the user is `absorb`ed, and an output will be `squeeze`d
+from the data structure. `digest` is a convenience method calling `absorb` and `squeeze` in one go. The `*_into`
+versions of these methods are for passing a buffer into which the calculated hashes are written.  The internal state
+will not be cleared unless `reset` is called.
+
+### Design of `CurlP`
+
+`CurlP` is designed as a hash function that acts on a `T1B1` binary-encoded ternary buffer, with a hash length of `243`
```

Comment:
consistency: in 76 it's *binary-coded*, here it's called *binary-encoded*

Reply:
Which one do you prefer ?

Reply:
To me the whole term is confusing, but I know that it's being used in ternary science, so we need to use it. Then I think I'ld prefer *binary-encoded*. :slightly_smiling_face: